File: 백준/Gold/15683. 감시/감시.py
import sys
import math
import copy


def input(): return sys.stdin.readline().rstrip()


N, M = map(int, input().split())


board = [list(map(int, input().split())) for _ in range(N)]

dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

cctv_mode = [
    [],
    [[0], [1], [2], [3]],
    [[0, 1], [2, 3]],
    [[0, 2], [0, 3], [1, 2], [1, 3]],
    [[0, 1, 2], [0, 1, 3], [0, 2, 3], [1, 2, 3]],
    [[0, 1, 2, 3]]
]


cctv = []
result = math.inf

# cctv 설치
for i in range(len(board)):
    for j in range(len(board[i])):
        if 1 <= board[i][j] <= 5:
            cctv.append((i, j, board[i][j]))


# 감시
def check(x, y, mode, board):
    for i in mode:
        nx, ny = x, y
        while True:
            nx += dx[i]
            ny += dy[i]

            if nx < 0 or ny < 0 or nx >= N or ny >= M:
                break
            if board[nx][ny] == 6:
                break
            if board[nx][ny] == 0:
                board[nx][ny] = -1


# 사각지대 개수
def cnt_zero(board):
    cnt = 0
    for i in board:
        cnt += i.count(0)
    return cnt


# 탐색
def dfs(depth, board):
    global result
    if depth == len(cctv):
        result = min(result, cnt_zero(board))
        return

    x, y, mode = cctv[depth]

    for i in cctv_mode[mode]:
        tmp_board = copy.deepcopy(board)
        check(x, y, i, tmp_board)
        dfs(depth + 1, tmp_board)
        tmp_board = copy.deepcopy(board)


dfs(0, board)
print(result)

File: 백준/Gold/15683. 감시/감시.java
import java.io.*;
import java.util.*;

public class Main {
    static int N, M, ans = Integer.MAX_VALUE;
    static int[][] office;
    static ArrayList<Camera> cameras = new ArrayList<>();

    static int[] dx = { -1, 0, 1, 0 };
    static int[] dy = { 0, 1, 0, -1 };

    static int[][][] directions = {
            {},
            { { 0 }, { 1 }, { 2 }, { 3 } },
            { { 0, 2 }, { 1, 3 } },
            { { 0, 1 }, { 1, 2 }, { 2, 3 }, { 3, 0 } },
            { { 0, 1, 2 }, { 1, 2, 3 }, { 2, 3, 0 }, { 3, 0, 1 } },
            { { 0, 1, 2, 3 } }
    };

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        office = new int[N][M];

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < M; j++) {
                office[i][j] = Integer.parseInt(st.nextToken());
                if (office[i][j] >= 1 && office[i][j] <= 5) {
                    cameras.add(new Camera(i, j, office[i][j]));
                }
            }
        }

        dfs(0, office);

        bw.write(String.valueOf(ans));
        bw.flush();
        bw.close();
        br.close();
    }

    static void dfs(int idx, int[][] board) {
        if (idx == cameras.size()) {
            int cnt = countBlindSpots(board);
            ans = Math.min(ans, cnt);
            return;
        }

        Camera cam = cameras.get(idx);
        int type = cam.type;

        for (int[] dirs : directions[type]) {
            int[][] copy = copyBoard(board);

            for (int d : dirs) {
                mark(copy, cam.x, cam.y, d);
            }
            dfs(idx + 1, copy);
        }
    }

    static int countBlindSpots(int[][] board) {
        int count = 0;
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (board[i][j] == 0)
                    count++;
            }
        }
        return count;
    }

    static int[][] copyBoard(int[][] board) {
        int[][] newBoard = new int[N][M];
        for (int i = 0; i < N; i++) {
            newBoard[i] = board[i].clone();
        }
        return newBoard;
    }

    static void mark(int[][] board, int x, int y, int d) {
        int nx = x + dx[d];
        int ny = y + dy[d];
        while (nx >= 0 && nx < N && ny >= 0 && ny < M && board[nx][ny] != 6) {

            if (board[nx][ny] == 0) {
                board[nx][ny] = -1;
            }

            nx += dx[d];
            ny += dy[d];
        }
    }

    static class Camera {
        int x, y, type;

        public Camera(int x, int y, int type) {
            this.x = x;
            this.y = y;
            this.type = type;
        }
    }
}